src/ray/raylet/scheduling_queue.cc

```diff
@@ -6,49 +6,90 @@ namespace ray {
 
 namespace raylet {
 
+bool SchedulingQueue::TaskQueue::Append(const TaskID &task_id, const Task &task) {
+  if (map_.find(task_id) != map_.end()) {
+    // This shouldn't happen.
+    return false;
+  }
+
+  auto list_iter = list_.insert(list_.end(), task);
+  map_[task_id] = list_iter;
+  return true;
+}
+
+bool SchedulingQueue::TaskQueue::Remove(const TaskID &task_id) {
+  auto iter = map_.find(task_id);
+  if (iter == map_.end()) {
+    // not exist.
+    return false;
+  }
+
+  auto list_iter = iter->second;
+  map_.erase(iter);
+  list_.erase(list_iter);
+  return true;
+}
+
+bool SchedulingQueue::TaskQueue::Remove(const TaskID &task_id, std::vector<Task> &removed_tasks) {
+  auto iter = map_.find(task_id);
+  if (iter == map_.end()) {
+    // not exist.
+    return false;
+  }
+
+  auto list_iter = iter->second;
+  removed_tasks.push_back(std::move(*list_iter));
+  map_.erase(iter);
```

Comment:
is it safe to erase after std::move on this element (e.g., should we be worried about a double free here)?

Reply:
It should be safe here. After ```std::move``` the specific task object as specified by ```*list_iter``` is not referenced any more, except for destruction, which is safe.  The ```erase``` basically removes the list iterator from hash, which is not impacted by std::move.